packages/server/src/restaurants/dishes/domain/UpdateDishHandler.ts

```diff
@@ -0,0 +1,44 @@
+import { Injectable } from '@nestjs/common';
+import { InjectModel } from '@nestjs/mongoose';
+import _ from 'lodash';
+import { Model } from 'mongoose';
+
+import { Handler } from '../../../shared';
+import { Restaurant, RestaurantDocument } from '../../database';
+import { Allergens, Dish, DishDocument, DishTags, Ingredient, MealType, NutritionalValue } from '../database';
+
+interface UpdateDishRequest {
+  readonly dishId: string;
+  readonly restaurantId: string;
+  readonly name: string;
+  readonly mealType: MealType[];
+  readonly description: string;
+  readonly price: number;
+  readonly tags: DishTags[];
+  readonly ingredients: Ingredient[];
+  readonly allergens: Allergens[];
+  readonly portionWeight: number;
+  readonly calories: NutritionalValue;
+  readonly fats: NutritionalValue;
+  readonly proteins: NutritionalValue;
+  readonly carbohydrates: NutritionalValue;
+}
+type sth = Omit<UpdateDishRequest, 'restaurantId' | 'dishId'>;
+@Injectable()
+class UpdateDishHandler implements Handler<UpdateDishRequest, Dish | undefined | null> {
+  constructor(
+    @InjectModel(Dish.name) private dishModel: Model<DishDocument>,
+    @InjectModel(Restaurant.name) private restaurantModel: Model<RestaurantDocument>,
+  ) {}
+  async exec(req: UpdateDishRequest): Promise<Dish | undefined | null> {
+    await this.restaurantModel.updateOne({ _id: req.restaurantId }, { $pull: { dishes: req.dishId } });
+    const result = await this.dishModel.findOneAndUpdate(
+      { _id: req.dishId, restaurant: req.restaurantId },
+      _.omit(req, 'dishId', 'restaurantId'),
+    );
+    if (result === null) return null;
```

Comment:
Widzę tu kilka problemów:
1.	`this.restaurantModel.updateOne` usuwa referencję do dish z restaurant, ale już nie ma drugiego push, który by dodawał zupdatowany dish. W konsekwencji update z poziomu restauracji działa jak delete.
2.	Jak już wcześniej ustalaliśmy, chcemy przechowywać historyczne wersje dish dla celów wyświetlania dish w już zrealizowanych orders. Wykonując `this.dishModel.findOneAndUpdate` podmienia się istniejącą wersję – tracąc historię.


Moim zdaniem flow powinno wyglądać tak (pomijając rollback/transakcje/failure recovery – tym się nie zajmujemy):
1.	Usunąć dishId z restaurant.dishes. W przypadku 0 zmodyfikowanych dokumentów ( nie znaleziono dishId w restaurant.dishes ) zwrócić 404
2.	Stworzyć nowe dish, które jest tak naprawdę updatem starego.
4.	Zaznaczyć w starym dish (req.dishId), że nie jest już aktualną wersją np. za pomocą flagi `updated: true` lub `replacedBy: newDish._id`.
5.	Dodać nowy dish do restaurant.dishes
6.	Zwrócić nowy dish, bo przecież zmieniło się id


Jest jeszcze druga opcja, zrobić kopię dish z nowym id, następnie we wszystkich miejscach, które trzymają referencję do dishId, zamienić dishId na dishId nowej kopi, następnie zrobić inplace update na bazowym dish. To rozwiązanie pozwala nam zachować to samo Id i wrócić 204, ale wymaga większej ilości updatów, ponieważ potencjalnie jest dużo więcej relacji dish-order niż restaurant-dish.